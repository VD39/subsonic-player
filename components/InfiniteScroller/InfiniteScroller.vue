<script setup lang="ts">
import ScrollerLoader from '@/components/Loaders/ScrollerLoader.vue';

const emit = defineEmits(['loadMore']);

emit('loadMore');

const { loading, hasMore } = useInfinityLoading();

const title = computed(() => (loading.value ? 'Loading data' : 'Load more'));

const container = ref<HTMLElement | null>(null);

function handleScroll() {
  if (
    !loading.value &&
    container.value!.getBoundingClientRect().bottom < window.innerHeight
  ) {
    emit('loadMore');
  }
}

onMounted(() => {
  if (container.value) {
    window.addEventListener('scroll', handleScroll);
  }
});

onUnmounted(() => {
  window.removeEventListener('scroll', handleScroll);
});

function loadMore() {
  console.log('loadMore');
  emit('loadMore');
}
</script>

<template>
  <div ref="container">
    <button
      v-if="hasMore"
      ref="button"
      :class="$style.button"
      type="button"
      :title="title"
      :disabled="loading"
      @click="loadMore"
    >
      <span v-if="!loading">Load more</span>

      <ScrollerLoader v-if="loading" />
    </button>
  </div>
</template>

<style module>
.button {
  @mixin align-center;

  justify-content: center;
  width: 150px;
  height: 50px;
  padding: var(--space-8) var(--space-12);
  margin: var(--space-32) auto;
  cursor: pointer;
  background-color: var(--dark-background-color);
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-m);

  &:disabled {
    pointer-events: none;
    opacity: 0.25;
  }
}
</style>
