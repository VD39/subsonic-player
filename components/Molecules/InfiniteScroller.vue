<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';

const emit = defineEmits(['loadMore']);

const { hasMore, loading } = useInfinityLoading();

const containerRef = ref<HTMLElement | null>(null);

const title = computed(() => (loading.value ? 'Loading data' : 'Load more'));

function loadMore() {
  emit('loadMore');
}

function onScroll() {
  if (
    !loading.value &&
    containerRef.value!.getBoundingClientRect().bottom < window.innerHeight
  ) {
    loadMore();
  }
}

loadMore();

onMounted(() => {
  if (!containerRef.value) {
    return;
  }

  window.addEventListener('scroll', onScroll);
});

onUnmounted(() => {
  window.removeEventListener('scroll', onScroll);
});
</script>

<template>
  <div ref="containerRef">
    <ButtonLink
      v-if="hasMore"
      :class="['centerAll', $style.buttonLink]"
      :title="title"
      :disabled="loading"
      show-text
      @click="loadMore"
    >
      <SpinningLoader v-if="loading" />

      <span v-else>Load more</span>
    </ButtonLink>

    <p v-else ref="message" :class="['centerAll', $style.message]">
      -- No more content --
    </p>
  </div>
</template>

<style module>
.buttonLink {
  width: 150px;
  height: 50px;
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-medium);
}

.buttonLink,
.message {
  margin: var(--space-24) auto 0;
}
</style>
