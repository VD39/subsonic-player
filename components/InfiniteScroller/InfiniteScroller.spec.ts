import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import InfiniteScroller from './InfiniteScroller.vue';

const windowAddEventListenerSpy = vi.spyOn(window, 'addEventListener');
const removeEventListenerSpy = vi.spyOn(window, 'removeEventListener');

const loadingMock = ref(false);
const hasMoreMock = ref(false);

mockNuxtImport('useInfinityLoading', () => () => ({
  loading: loadingMock,
  hasMore: hasMoreMock,
}));

function factory(props = {}) {
  return mount(InfiniteScroller, {
    props: {
      ...props,
    },
    attachTo: document.body,
  });
}

describe('InfiniteScroller', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  afterEach(() => {
    loadingMock.value = false;
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('emits the loadMore event', () => {
    expect(wrapper.emitted('loadMore')).toEqual([[]]);
  });

  it('adds the scroll event listener function', () => {
    expect(windowAddEventListenerSpy).toHaveBeenCalledWith(
      'scroll',
      expect.any(Function),
    );
  });

  describe('when hasMore value is false', () => {
    it('does not show the button element', () => {
      expect(wrapper.find({ ref: 'button' }).exists()).toBe(false);
    });
  });

  describe('when hasMore value is true', () => {
    beforeEach(() => {
      hasMoreMock.value = true;
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the button element', () => {
      expect(wrapper.find({ ref: 'button' }).exists()).toBe(true);
    });

    describe('when loading value is false', () => {
      it('sets the correct title value', () => {
        expect(wrapper.find({ ref: 'button' }).attributes('title')).toBe(
          'Load more',
        );
      });
    });

    describe('when loading value is true', () => {
      beforeEach(() => {
        loadingMock.value = true;
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('sets the correct title value', () => {
        expect(wrapper.find({ ref: 'button' }).attributes('title')).toBe(
          'Loading data',
        );
      });
    });

    describe('when load more button is clicked', () => {
      beforeEach(async () => {
        await wrapper.find({ ref: 'button' }).trigger('click');
      });

      it('calls the loadMore event', () => {
        expect(wrapper.emitted('loadMore')).toEqual([[], []]);
      });
    });

    describe('when scroll is called', () => {
      beforeEach(() => {
        global.window.innerHeight = 500;
        Element.prototype.getBoundingClientRect = vi.fn(
          () =>
            ({
              bottom: 500,
            }) as DOMRect,
        );
      });

      describe('when loading value is true', () => {
        it('does not emit the loadMore event', () => {
          expect(wrapper.emitted('loadMore')).toEqual([[]]);
        });
      });

      describe('when loading value is false', () => {
        beforeEach(() => {
          loadingMock.value = false;
          global.dispatchEvent(new CustomEvent('scroll'));
        });

        describe('when innerHeight is less than container getBoundingClientRect bottom', () => {
          beforeEach(() => {
            global.dispatchEvent(new CustomEvent('scroll'));
          });

          it('does not emit the loadMore event', () => {
            expect(wrapper.emitted('loadMore')).toEqual([[]]);
          });
        });

        describe('when innerHeight is greater than container getBoundingClientRect bottom', () => {
          beforeEach(() => {
            global.window.innerHeight = 600;
            global.dispatchEvent(new CustomEvent('scroll'));
          });

          it('emits the loadMore event', () => {
            expect(wrapper.emitted('loadMore')).toEqual([[], []]);
          });
        });
      });
    });

    describe('when component unmounts', () => {
      beforeEach(() => {
        wrapper.unmount();
      });

      it('removes the scroll event listeners function', () => {
        expect(removeEventListenerSpy).toHaveBeenCalledWith(
          'scroll',
          expect.any(Function),
        );
      });
    });
  });
});
