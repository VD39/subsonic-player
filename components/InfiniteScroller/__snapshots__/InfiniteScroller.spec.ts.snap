// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfiniteScroller > matches the snapshot 1`] = `
"<div>
  <p class="message">-- No more content --</p>
</div>"
`;

exports[`InfiniteScroller > when hasMore value is true > matches the snapshot 1`] = `
"<div><button class="button" type="button" title="Load more"><span>Load more</span>
    <!--v-if-->
  </button></div>"
`;

exports[`InfiniteScroller > when hasMore value is true > when loading value is true > matches the snapshot 1`] = `
"<div><button class="button" type="button" title="Loading data" disabled="">
    <!--v-if--><span><span class="loader"></span><span class="visually-hidden">Loading data</span></span>
  </button></div>"
`;
